src/test/app/Credentials_test.cpp

```diff
@@ -0,0 +1,770 @@
+                    "asdfghjkl;'zxcvbnm8237tr28weufwldebvfv8734t07p   "
+                    "9hfup;wDJFBVSD8f72  "
+                    "pfhiusdovnbs;djvbldafghwpEFHdjfaidfgio84763tfysgdvhjasbd "
+                    "vujhgWQIE7F6WEUYFGWUKEYFVQW87FGWOEFWEFUYWVEF8723GFWEFBWULE"
+                    "fv28o37gfwEFB3872TFO8GSDSDVD";
+                static_assert(longCredType.size() > maxCredentialTypeLength);
+                auto jv = credentials::create(subject, issuer, longCredType);
+                env(jv, ter(temMALFORMED));
+            }
+
+            {
+                testcase("Credentials fail, URI length > 256.");
+                constexpr std::string_view longURI =
+                    "abcdefghijklmnopqrstuvwxyz01234567890qwertyuiop[]"
+                    "asdfghjkl;'zxcvbnm8237tr28weufwldebvfv8734t07p   "
+                    "9hfup;wDJFBVSD8f72  "
+                    "pfhiusdovnbs;djvbldafghwpEFHdjfaidfgio84763tfysgdvhjasbd "
+                    "vujhgWQIE7F6WEUYFGWUKEYFVQW87FGWOEFWEFUYWVEF8723GFWEFBWULE"
+                    "fv28o37gfwEFB3872TFO8GSDSDVD";
+                static_assert(longURI.size() > maxCredentialURILength);
+                env(credentials::create(subject, issuer, credType),
+                    credentials::uri(longURI),
+                    ter(temMALFORMED));
+            }
+
+            {
+                testcase("Credentials fail, URI empty.");
+                env(credentials::create(subject, issuer, credType),
+                    credentials::uri(""),
+                    ter(temMALFORMED));
+            }
+
+            {
+                testcase("Credentials fail, expiration in the past.");
+                auto jv = credentials::create(subject, issuer, credType);
+                // current time in ripple epoch - 1s
+                uint32_t const t = env.now().time_since_epoch().count() - 1;
+                jv[sfExpiration.jsonName] = t;
+                env(jv, ter(tecEXPIRED));
+            }
+
+            {
+                testcase("Credentials fail, invalid fee.");
+
+                auto jv = credentials::create(subject, issuer, credType);
+                jv[jss::Fee] = -1;
+                env(jv, ter(temBAD_FEE));
+            }
+
+            {
+                testcase("Credentials fail, duplicate.");
+                auto const jv = credentials::create(subject, issuer, credType);
+                env(jv);
+                env.close();
+                env(jv, ter(tecDUPLICATE));
+                env.close();
+            }
+        }
+
+        {
+            using namespace jtx;
+            Env env{*this, features};
+
+            env.fund(XRP(5000), issuer);
+            env.close();
+
+            {
+                testcase("Credentials fail, subject doesn't exist.");
+                auto const jv = credentials::create(subject, issuer, credType);
+                env(jv, ter(tecNO_TARGET));
+            }
```

Comment:
This testcase could be simplified (no need to re-fund the issuer) if you create a `subject2` account that you use for this test.

Reply:
It is simple enough, to simplify it more